ambry-store/src/main/java/com/github/ambry/store/BlobStore.java

```diff
@@ -801,8 +803,7 @@ public short undelete(MessageInfo info) throws StoreException {
           FileSpan fileSpan = new FileSpan(indexEndOffsetBeforeCheck, currentIndexEndOffset);
           IndexValue value = index.findKey(info.getStoreKey(), fileSpan,
               EnumSet.of(PersistentIndex.IndexEntryType.DELETE, PersistentIndex.IndexEntryType.UNDELETE));
-          if (value != null && value.getOffset().compareTo(indexEndOffsetBeforeCheck) >= 0) {
-            // Make sure the value is actually after the indexEndOffsetBeforeCheck
+          if (value != null) {
             if (value.isUndelete() && value.getLifeVersion() == revisedLifeVersion) {
```

Comment:
I need a little clarification of this if-else block. The `if` branch makes sense to me but for the `else` branch, what if the `value.isDelete() == true`,  it seems to throw exception directly.  Actually this might be a valid case. 
(Correct me if I am wrong and let me know anything I missed here)

Reply:
It's not a valid delete. If we are here, then the previous record should be DELETE. For example, if the index values for this blob is PUT, DELETEv0. Then this UNDELETE's lifeVersion should be 1. so in `if` statement, we see a UNDELETEv1 out of the previous scope, this UNDELETEv1 might come from replication, this is valid. But if it's not UNDELETEv1, then all the alternatives should be invalid. For instance, UNDELETEv2 is invalid since when the undelete method is invoked, the DELETE is at version 0. DELETEv2 is also invalid since we don't really know the order of DELETEv2 and UNDELETEv1 anymore. Imagine that this host some how freezes for a long time, and within this period, client send a DELETE, UNDELETE and DELETE. So in other host, the final state is DELETE, if we append UNDELETE after this DELETE, then we change the final state. This is not right.

Basically what I am doing here is to make sure that only same UNDELETE from replication can be valid.